### PR TITLE
ISPN-5019 After coordinator change, cache topologies should be installed in parallel

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-5019

* Update LargeClusterStressTest to measure the cluster shutdown as well
* Improve the memory usage and serialization of consistent hash classes
* Install merge cache topologies in parallel
* Don't keep the state transfer thread blocked while waiting for state
* Don't use up a thread for early StateRequestCommands
* Add a JGroupsAddress cache

### CODE REVIEW

File: commons/src/main/java/org/infinispan/commons/marshall/Ids.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
    int EMPTY_MAP = 89;
    int EMPTY_LIST = 90;
    // internal collections (id=18 no longer in use, might get reused at a later stage)
+   int IMMUTABLE_LIST = 18;
```
Comment (tristantarrant) [inline commons/src/main/java/org/infinispan/commons/marshall/Ids.java]: I guess the comment above doesn't make sense anymore :)

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 2]
```diff
@@ -518,13 +544,13 @@ private void applyChunk(Address sender, Set<Integer> mySegments, StateChunk stat
...
    private void doApplyState(Address sender, int segmentId, Collection<InternalCacheEntry> cacheEntries) {
-      log.debugf("Applying new state chunk for segment %d of cache %s from node %s: received %d cache entries", segmentId, cacheName, sender, cacheEntries.size());
+      log.tracef("Applying new state chunk for segment %d of cache %s from node %s: received %d cache entries",
+            segmentId, cacheName, sender, cacheEntries.size());
       if (trace) {
          List<Object> keys = new ArrayList<Object>(cacheEntries.size());
          for (InternalCacheEntry e : cacheEntries) {
             keys.add(e.getKey());
          }
-         log.tracef("Received chunk with keys %s for segment %d of cache %s from node %s", keys, segmentId, cacheName, sender);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: well, if you remove the trace message, you can also remove the full ```if``` :)

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 3]
```diff
@@ -657,7 +685,7 @@ private void addTransfers(Set<Integer> segments) {
          requestSegments(segments, sources, excludedSources);
       }
 
-      log.debugf("Finished adding inbound state transfer for segments %s of cache %s", segments, cacheName);
+      log.tracef("Finished adding inbound state transfer for segments %s of cache %s", segments, cacheName);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: ```if (trace) ...```

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 4]
```diff
@@ -1005,8 +982,11 @@ private InboundTransferTask addTransfer(Address source, Set<Integer> segmentsFro
       synchronized (transferMapsLock) {
          log.tracef("Adding transfer from %s for segments %s", source, segmentsFromSource);
          segmentsFromSource.removeAll(transfersBySegment.keySet());  // already in progress segments are excluded
-         if (!segmentsFromSource.isEmpty()) {
-            InboundTransferTask inboundTransfer = new InboundTransferTask(segmentsFromSource, source,
+         if (segmentsFromSource.isEmpty()) {
+            log.tracef("All segments are already in progress, skipping");
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: ```if (trace)```

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 5]
```diff
@@ -255,9 +256,9 @@ private boolean updateCacheTopology(String cacheName, CacheTopology cacheTopolog
          } else {
             log.debugf("Updating local topology for cache %s: %s", cacheName, cacheTopology);
             cacheStatus.setCurrentTopology(cacheTopology);
-            return true;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: is this change really needed?

File: core/src/test/java/org/infinispan/test/TestingUtil.java [hunk 6]
```diff
@@ -695,8 +698,8 @@ public static void clearContent(EmbeddedCacheManager cacheContainer) {
          }
       }
 
-      if (cacheContainer.isDefaultRunning()) {
-         Cache defaultCache = cacheContainer.getCache();
+      Cache defaultCache = cacheContainer.getCache(EmbeddedCacheManager.DEFAULT_CACHE_NAME);
+      if (defaultCache != null) {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/test/TestingUtil.java]: I think we never return ```null```

File: commons/src/main/java/org/infinispan/commons/marshall/Ids.java [hunk 7]
```diff
@@ -14,7 +14,8 @@
    int EMPTY_SET = 88;
    int EMPTY_MAP = 89;
    int EMPTY_LIST = 90;
-   // internal collections (id=18 no longer in use, might get reused at a later stage)
+   
+   int IMMUTABLE_LIST = 18;
```
Comment (pruivo) [inline commons/src/main/java/org/infinispan/commons/marshall/Ids.java]: why creating a new ```Externalizer``` instead of updating the ```ListExternalizer```?

Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/marshall/Ids.java]: The immutable list is quite different than the others, because it must create an array and wrap it instead of using `add(element)`.

File: core/src/test/java/org/infinispan/test/TestingUtil.java (hunk 6 above)
Comment (danberindei) [inline core/src/test/java/org/infinispan/test/TestingUtil.java]: We do return null if there was an exception during `cache.start()`, because we insert the `CacheWrapper` in the map of caches but we don't call `setCache`.

I did make a mistake here though, I should have kept the `isDefaultRunning()` check, otherwise it could start the default cache.

File: core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same time.
+ *
+ * @author Dan Berindei
+ * @since 7.2
+ */
+public class SemaphoreCompletionService<T> implements CompletionService<T> {
+   private static final Log log = LogFactory.getLog(SemaphoreCompletionService.class);
+   // Trace logs are always disabled by default
+   private static final boolean trace = false;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java]: Did you mean for this to be log.isTraceEnabled() ?

File: core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+      for (QueueingTask task : queuedTasks) {
+         task.cancel(false);
+      }
+   }
+
+   /**
+    * Called from a task to keep a thread out-of-reach for other tasks.
+    */
+   public void continueTaskInBackground() {
+      if (trace) log.tracef("Moving task to background, available permits %d", semaphore.availablePermits());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java]: Did we want this to log after we freed the permit ?

File: core/src/main/java/org/infinispan/marshall/exts/ListExternalizer.java [hunk 10]
```diff
@@ -47,17 +47,18 @@ public void writeObject(ObjectOutput output, List list) throws IOException {
    @Override
    public List readObject(ObjectInput input) throws IOException, ClassNotFoundException {
       int magicNumber = input.readUnsignedByte();
+      int size = UnsignedNumeric.readUnsignedInt(input);
+
       List<Object> subject = null;
       switch (magicNumber) {
          case ARRAY_LIST:
-            subject = new ArrayList<Object>();
+            subject = new ArrayList<Object>(size);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/exts/ListExternalizer.java]: :+1:

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+ * Cache JGroupsAddress instances
+ *
+ * @author Dan Berindei
+ * @since 7.0
+ */
+public class JGroupsAddressCache {
+   private static final EquivalentConcurrentHashMapV8<Address, JGroupsAddress> addressCache =
+         new EquivalentConcurrentHashMapV8<>(AnyEquivalence.getInstance(), AnyEquivalence.getInstance());
+
+   static org.infinispan.remoting.transport.Address fromJGroupsAddress(final org.jgroups.Address addr) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java]: Why not just return JGroupsAddress ?

Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java]: By the way did you see a big improvement in overall message size with the cache ?

Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java]: Also I wonder we could have 2 methods here, 1 that takes a regular org.jgroups.Address and another that takes a org.jgroups.util.ExtendedUUID and returns the proper type.  At least then the cast is isolated in 1 spot and not every caller has to cast like we have atm.

File: commons/src/main/java/org/infinispan/commons/marshall/Ids.java (hunk 7 above)
Comment (pruivo) [inline commons/src/main/java/org/infinispan/commons/marshall/Ids.java]: yeah, but the change is small enough that can be done in the switch-case branch of ```ListExternalizer```

Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/marshall/Ids.java]: Only if I also put a `return` in the switch case, which I really don't like :)

File: core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java (hunk 8 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java]: I was thinking none of the existing executors log anything, so I'd make this the same. But I see `BlockingTaskAwareExecutorImpl` does have some trace logs, so I'll change it.

File: core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java (hunk 9 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java]: Moving the task in background doesn't free a permit, that's the point :)

File: core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java (hunk 8 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/executors/SemaphoreCompletionService.java]: Yeah this should only be evaluated once since it is static so I see no reason not to.

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java (hunk 11 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java]: I guess you don't like the return idea ?

Comment (danberindei) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java]: 1) I think I did it because I thought JGroupsTopologyAwareAddress didn't extend JGroupsAddress, but anyway I like returning an interface better.
2) I didn't check the message size, but the memory usage with 40 nodes x 80 caches improved quite a bit, and so did the performance.
3) I only expect the `instanceof` check to be performed about for each other node, so it shouldn't matter.

**Comment by danberindei** [comment]
Updated to remove some integrationtests/security-it debug crutches.

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java [hunk 12]
```diff
@@ -66,6 +52,20 @@
 import org.infinispan.util.logging.Log;
 import org.infinispan.util.logging.LogFactory;
 
+import java.io.IOException;
```
Comment (galderz) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: Moving around imports? :wink:

File: core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java [hunk 13]
```diff
@@ -116,6 +109,14 @@
 import org.jboss.marshalling.ObjectTable;
 import org.jboss.marshalling.Unmarshaller;
 
+import java.io.IOException;
```
Comment (galderz) [inline core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java]: And again....

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddress.java [hunk 14]
```diff
@@ -1,15 +1,15 @@
...
-import java.io.ObjectInput;
-import java.io.ObjectOutput;
-import java.util.Set;
-
 import org.infinispan.commons.marshall.InstanceReusingAdvancedExternalizer;
 import org.infinispan.commons.util.Util;
 import org.infinispan.marshall.core.Ids;
 import org.infinispan.remoting.transport.Address;
 
+import java.io.IOException;
```
Comment (galderz) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddress.java]: More useless import reshuffling...

**Comment by galderz** [comment]
Looks good, integrating. Please leave import reshuffling for another time ;)

**Comment by galderz** [comment]
Integrated

**Comment by danberindei** [comment]
Thanks, Galder. I'm not sure why we have all this import shuffling, I thought the code style settings also define a fixed order for imports?

**Comment by wburns** [comment]
@danberindei afaik it is because some people use Eclipse and some use IntelliJ.  For whatever reason the defaults for standard java classes is in different spots on each.  I would have thought this could have been defined as well in the  style config.

**Comment by danberindei** [comment]
@wburns Oh well... in the meantime I found the "optimize imports on the fly" setting in intellij and I disabled it :)

**Comment by wburns** [comment]
@danberindei ah nice, Eclipse should have something like that then too :)
